Amazon.QLDB.Driver.Tests/AsyncRetryHandlerTests.cs

```diff
@@ -0,0 +1,132 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
+ * the License. A copy of the License is located at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+namespace Amazon.QLDB.Driver.Tests
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Net;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Amazon.QLDBSession;
+    using Amazon.QLDBSession.Model;
+    using Amazon.Runtime;
+    using Microsoft.Extensions.Logging.Abstractions;
+    using Microsoft.VisualStudio.TestTools.UnitTesting;
+    using Moq;
+
+    [TestClass]
+    public class AsyncRetryHandlerTests
+    {
+        [DataTestMethod]
+        [DynamicData(nameof(CreateRetriableExecuteAsyncTestData), DynamicDataSourceType.Method)]
+        public async Task RetriableExecuteAsync_RetryOnExceptions(Driver.RetryPolicy policy, IList<Exception> exceptions, Type expectedExceptionType, Type innerExceptionType,
+            Times funcCalledTimes, Times newSessionCalledTimes, Times nextSessionCalledTimes)
+        {
+            var handler = (AsyncRetryHandler)AsyncQldbDriverBuilder.CreateDefaultRetryHandler(NullLogger.Instance);
+
+            var func = new Mock<Func<CancellationToken, Task<int>>>();
+            var newSession = new Mock<Func<CancellationToken, Task>>();
+            var nextSession = new Mock<Func<CancellationToken, Task>>();
+
+            var seq = func.SetupSequence(f => f.Invoke(default));
+            foreach (var ex in exceptions)
+            {
+                seq = seq.ThrowsAsync(ex);
+            }
+            seq.ReturnsAsync(1);
+
+            try
+            {
+                await handler.RetriableExecute(
+                    func.Object,
+                    policy,
+                    newSession.Object,
+                    nextSession.Object);
+
+                Assert.IsNull(expectedExceptionType);
+            }
+            catch (Exception e)
+            {
+                Assert.IsNotNull(expectedExceptionType);
+                Assert.IsInstanceOfType(e, expectedExceptionType);
+
+                if (innerExceptionType != null)
+                {
+                    Assert.IsInstanceOfType(e.InnerException, innerExceptionType);
+                }
+            }
+
+            func.Verify(f => f.Invoke(default), funcCalledTimes);
+            newSession.Verify(r => r.Invoke(default), newSessionCalledTimes);
+            nextSession.Verify(r => r.Invoke(default), nextSessionCalledTimes);
+        }
+
+        public static IEnumerable<object[]> CreateRetriableExecuteAsyncTestData()
```

Comment:
It is possible to somehow use share this between `AsyncRetryHandlerTests` and `RetryHandlerTests`?

Reply:
Technically this one returns one fewer field than `RetryHandlerTests` because we're not testing the retry action here.

Reply:
If we can do something like:
`[TestMethod]
 [DynamicData(nameof(CreateRetriableExecuteAsyncTestData), DynamicDataSourceType.Method)]
 [CreateRetriableExecuteAsyncTestData]`
 
 To solve the issue with having a dynamicData in another file, would it be worthwhile to have this method only once, and then here remove that extra field? It's around 50 lines of duplicated code

Reply:
Well if we can't define dynamic data in another file then we have no choice but to duplicate the definition.